modules/cosmosBidAdapter.js

```diff
@@ -0,0 +1,398 @@
+import { registerBidder } from '../src/adapters/bidderFactory';
+import { BANNER, VIDEO } from '../src/mediaTypes';
+import * as utils from '../src/utils';
+
+const BIDDER_CODE = 'cosmos';
+const BID_ENDPOINT = '//bid.cosmoshq.com/openrtb2/bids';
+const USER_SYNC_ENDPOINT = '//sync.cosmoshq.com/js/v1/usersync.html';
+const HTTP_POST = 'POST';
+const LOG_PREFIX = 'COSMOS: ';
+const DEFAULT_CURRENCY = 'USD';
+const HTTPS = 'https:';
+const MEDIA_TYPES = 'mediaTypes';
+const DEFAULT_NET_REVENUE = false;
+
+export const spec = {
+  code: BIDDER_CODE,
+  supportedMediaTypes: [BANNER, VIDEO],
+
+  /**
+   * generate UUID
+   **/
+  _createUUID: function () {
+    var u = Date.now().toString(16) + Math.random().toString(16) + '0'.repeat(16);
+    return [u.substr(0, 8), u.substr(8, 4), '4000-8' + u.substr(13, 3), u.substr(16, 12)].join('-');
+  },
+
+  /**
+   * copy object if not null
+   **/
+  _copyObject: function (src, dst) {
+    if (src) {
+      // copy complete object
+      Object.keys(src).forEach(param => dst[param] = src[param]);
```

Comment:
I believe this type of copying could leave a reference of the original object in the copied object, which could mean if the original object was modified  - the copied object would mutate and use the new value(s).

There is a `utils.deepClone()` function available if you want to safely copy one object to another.

Reply:
Used utils.deepClone() in buildRequests function to clone `validBidRequests` input params from pbs lib. Need this function to give object preference as per cosmos logic.